lib/active_merchant/billing/gateways/pro_pay.rb

```diff
@@ -0,0 +1,327 @@
+        "4" => "Pick up card and deny transaction",
+        "5" => "Problem with the account",
+        "6" => "Customer requested stop to recurring payment",
+        "7" => "Customer requested stop to all recurring payments",
+        "8" => "Honor with ID only",
+        "9" => "Unpaid items on customer account",
+        "12" => "Invalid transaction",
+        "13" => "Amount Error",
+        "14" => "Invalid card number",
+        "15" => "No such issuer. Could not route transaction",
+        "16" => "Refund error",
+        "17" => "Over limit",
+        "19" => "Reenter transaction or the merchant account may be boarded incorrectly",
+        "25" => "Invalid terminal 41 Lost card",
+        "43" => "Stolen card",
+        "51" => "Insufficient funds",
+        "52" => "No such account",
+        "54" => "Expired card",
+        "55" => "Incorrect PIN",
+        "57" => "Bank does not allow this type of purchase",
+        "58" => "Credit card network does not allow this type of purchase for your merchant account.",
+        "61" => "Exceeds issuer withdrawal limit",
+        "62" => "Issuer does not allow this card to be charged for your business.",
+        "63" => "Security Violation",
+        "65" => "Activity limit exceeded",
+        "75" => "PIN tries exceeded",
+        "76" => "Unable to locate account",
+        "78" => "Account not recognized",
+        "80" => "Invalid Date",
+        "82" => "Invalid CVV2",
+        "83" => "Cannot verify the PIN",
+        "85" => "Service not supported for this card",
+        "93" => "Cannot complete transaction. Customer should call 800 number.",
+        "95" => "Misc Error Transaction failure",
+        "96" => "Issuer system malfunction or timeout.",
+        "97" => "Approved for a lesser amount. ProPay will not settle and consider this a decline.",
+        "98" => "Failure HV",
+        "99" => "Generic decline or unable to parse issuer response code"
+      }
+
+      def initialize(options={})
+        requires!(options, :cert_str)
+        super
+      end
+
+      def purchase(money, payment, options={})
+        request = build_xml_request do |xml|
+          add_invoice(xml, money, options)
+          add_payment(xml, payment, options)
+          add_address(xml, options)
+          add_account(xml, options)
+          add_recurring(xml, options)
+          xml.transType "04"
+        end
+
+        commit(request)
+      end
+
+      def authorize(money, payment, options={})
+        request = build_xml_request do |xml|
+          add_invoice(xml, money, options)
+          add_payment(xml, payment, options)
+          add_address(xml, options)
+          add_account(xml, options)
+          add_recurring(xml, options)
+          xml.transType "05"
+        end
+
+        commit(request)
+      end
+
+      def capture(money, authorization, options={})
+        request = build_xml_request do |xml|
+          xml.amount money if money
+          xml.transNum authorization
```

Comment:
Hmmm. Sometimes I think it's ok to have the method just add the single value instead of always using any extracted methods. But, the more I think about it, it's probably better for me to keep using to extracted methods.